src/sql/workbench/contrib/query/browser/messagePanel.ts

```diff
@@ -188,30 +196,53 @@ export class MessagePanel extends Disposable {
 	}
 
 	public set queryRunner(runner: QueryRunner) {
+		this.runner = runner;
+
 		if (this.currenturi) {
 			this._treeStates.set(this.currenturi, this.tree.getViewState());
 		}
 		this.queryRunnerDisposables.clear();
 		this.reset();
 		this.currenturi = runner.uri;
-		this.queryRunnerDisposables.add(runner.onQueryStart(() => this.reset()));
-		this.queryRunnerDisposables.add(runner.onMessage(e => this.onMessage(e)));
-		this.onMessage(runner.messages, true);
+
+		this.queryRunnerDisposables.add(runner.onQueryStart(() => {
+			this.model.messages = [];
+			this.model.totalExecuteMessage = undefined;
+			this.tree.updateChildren();
+
+			this.queryResultsWriter.onQueryStart();
+		}));
+
+		this.queryRunnerDisposables.add(runner.onMessage((message) => {
+			this.queryResultsWriter.onMessage(message);
+		}));
+
+		this.queryRunnerDisposables.add(runner.onResultSet((resultSet) => {
+			this.queryResultsWriter.onResultSet(resultSet);
+		}));
+
+		this.queryRunnerDisposables.add(runner.onResultSetUpdate((resultSet) => {
+			this.queryResultsWriter.updateResultSet(resultSet);
+		}));
+
+		if (this.queryResultsWriterStatus.isWritingToGrid()) {
```

Comment:
For all these checks - why is the writer itself not doing the filtering to determine whether it wants to actually write the stuff to its output? That would generally be preferred so we don't have that kind of logic spread out all over the application.